### DISCUSSION TITLE
Why are 2 `Settings`(config) loaded at all times?

### DISCUSSION THREAD

**Discussion by hasezoey** [tramhao/termusic#214] [Q&A] [answered]
I have just noticed that there are always 2 `Settings`(config) instances loaded at all times.
[Once in `server/src/server.rs`](https://github.com/tramhao/termusic/blob/f72757e829fa4a9c20ecce5e0cac21b2a830945a/server/src/server.rs#L41) and later [gets cloned over to `playback/src/lib.rs`'s `GeneralPlayer`](https://github.com/tramhao/termusic/blob/f72757e829fa4a9c20ecce5e0cac21b2a830945a/playback/src/lib.rs#L243).

This may cause de-syncs (and confusion) between those 2 instances, and they seem to be used interchangeably (and seemingly randomly), meaning some stuff uses the player's config to check values, while some other stuff checks the other config.

Unless there is a specific reason (which i would like to document), there should likely be only 1 `Settings`(config) loaded at all times to have 1 source of truth.

As for the fixes i would propose would be to have a shared instance via a `Arc<RwLock>` or `Arc<Mutex>`.

Note: with "at all times" i mean loaded from program beginning to end, not meaning temporary loads like loading a new config and validate it before using.

**Comment by tramhao** [comment]
Because originally it's only one app. Then it's divided to server and client. It's best to separate the config into server.toml and tui.toml. However, there are some configuration that are needed for both.

**Reply by hasezoey** [reply]
i agree, but that was not the question, the question is why there are 2 instances active inside *server*

**Reply by hasezoey** [reply]
actually now that i look at it closer, it looks like there are more than 2 instances of `Settings` being stored

**Comment by tramhao** [comment, marked as answer]
Yes, it's a problem. Should be centralized. Basically configs are referenced quite a lot. Sometimes when one option need to change, it's cloned. After change, it'll be saved. It's not a good way. When config get bigger, it's hard to manage.

**Reply by hasezoey** [reply]
i am currently refactoring it, will post a pr in a bit

**Comment by tramhao** [comment]
And there are more configs in TUI:-)

**Reply by hasezoey** [reply]
i am currently still working on the shared config refactor; i am finished with the server part, but it will take a bit longer for the TUI part.
do you want to wait until this is finished before you release the 0.8.0 or should it be included in the next release? (my estimate is i will be finished by tomorrow at the latest)

**Reply by tramhao** [reply]
Of course I'll wait, no rush. I met problem for new release.  #259

**Reply by hasezoey** [reply]
for reference, the PR is up as #260
